### PR TITLE
new smtp conf parameters: skip certs verification & custom root ca

### PR DESCRIPTION
**Pull request by KarolBedkowski**
Two new configuration parameters related to smtp with tls:
smtp_tls_skip_verify / tls_skip_verify - do not verify certificates
smtp_tls_root_ca / tls_root_ca - use custom certificate for smtp server from pem file

@stuartnelson3

### CODE REVIEW

**Comment by sgissi** [comment]
I have disabled TLS verification once because my SMTP provider would not
provide a correct SAN but it is in general a bad practice. I'm totally Ok
with custom CA in the other hand.

2017-04-28 9:41 GMT-07:00 Karol Będkowski <notifications@github.com>:

**Comment by KarolBedkowski** [comment]
I agree, but disabling verification is quite useful in broken/test environments - when other methods fail.

**Comment by KarolBedkowski** [comment]
Hi,
@stuartnelson3 - I've updated this path; now it use prometheus/common/config.TLSConfig for configuration.

**Comment by stuartnelson3** [comment]
I'll bounce this to @brian-brazil and @fabxc since they should be the ones making this decision

File: notify/impl.go [hunk 1]
```diff
@@ -272,7 +272,14 @@ func (n *Email) Notify(ctx context.Context, as ...*types.Alert) (bool, error) {
...
 		}
-		tlsConf := &tls.Config{ServerName: host}
+		tlsConf := &tls.Config{}
+		if n.conf.TLSConfig != nil {
+			tlsConf, err = n.conf.TLSConfig.GenerateConfig()
+			if err != nil {
+				return true, fmt.Errorf("generate TLS configuration failed: %s", err)
+			}
+		}
+		tlsConf.ServerName = host
```
Comment (brian-brazil) [inline notify/impl.go]: Why are you setting this?

File: config/config.go [hunk 2]
```diff
@@ -317,6 +321,8 @@ type GlobalConfig struct {
 	OpsGenieAPIHost  string `yaml:"opsgenie_api_host" json:"opsgenie_api_host"`
 	VictorOpsAPIURL  string `yaml:"victorops_api_url" json:"victorops_api_url"`
 
+	SMTPTLSConfig *config.TLSConfig `yaml:"smtp_tls_config"`
```
Comment (brian-brazil) [inline config/config.go]: I don't think a struct like this works as a default, all the rest are simple types. This should possibly be per-receiver only.

**Comment by KarolBedkowski** [comment]
@brian-brazil 
In my environment I have to set custom ca certificate for smtp server (or turn off whole certificate verification).
I don't want to setup this in per-receiver - I have one smtp server and many email receivers. Set it globally (like stmp_smarthost)  IMHO simplify configuration. Of course, change/set it in each receiver is possible (tls_config: ...) if you don't want setup it globally.  I think, this is ok for smtp and maybe webhook - I don't use other.

I've update patch - remove one unnecessary tls.Config allocation.

**Comment by brian-brazil** [comment]
This is something you could handle in your configuration management system.

**Comment by KarolBedkowski** [comment]
Unfortunately, I can't. My env is.. specific.
I need to use smtp directly, without smart host, I need accept custom ca but I can't add this to global system system certificates, etc.

**Comment by brian-brazil** [comment]
We generally presume that our users have configuration management that includes some form of templating system, as that's pretty much essential for managing any non-trivial environment.

**Comment by KarolBedkowski** [comment]
Well, there are also legacy system / corpo-world, where we can't change too much what we get and we try to survive.

Anyway, still Alertmanager can't talk with smtp when custom ca / client certificates are required. 

I propose simple patch for change this. It will be nice, if it be merged (I think, there are more people facing this problem). If not - no problem for me.

**Comment by JeffHemmen** [comment]
+1

**Comment by simonpasquier** [comment]
Closed by #1528

**Pull request closed without merging**